enhancements/network/on-prem-service-load-balancers.md

```diff
@@ -0,0 +1,405 @@
+---
+title: on-prem-service-load-balancers
+authors:
+  - "@russellb"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-05-14
+last-updated: 2020-05-14
+status: proposed
+---
+
+# Service Load Balancers for On Premise Infrastructure
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+We do not currently support [Services of
+type=LoadBalancer](https://kubernetes.io/docs/concepts/services-networking/#loadbalancer)
+(Service Load Balancers, or SLBs) for OpenShift in bare metal environments.
+While bare metal clusters are of primary interest, we hope to find a solution
+that would apply to other on-premise environments that don’t have native load
+balancer capabilities available (a cluster on VMware, RHV, or OpenStack, for
+example).
+
+## Motivation
+
+Service Load Balancers are a common way to expose applications on a cluster. We
+highly value clusters using on premise infrastructure, but do not support this
+feature in that context.  We aim to fill this gap with an optional range of
+capabilities.
+
+### Goals
+
+A SLB solution must provide these high level features:
+
+* Management of one or more pools of IP addresses to be allocated for SLBs
+* High Availability (HA) management of these IP addresses once allocated.  We
+  must be able to fail over addresses in less than 5 seconds.
+* Advertisement of the location of an IP address once assigned to a Node or
+  moved to a new Node.  Possible approaches include:
+    * Layer 2 (gratuitous ARP for IPv4, NDP for IPv6) - good for wide range of
+      environment compatibility, but limiting for larger clusters
+    * Layer 3 (BGP) - good for integration with networks for larger clusters
+* Suitable for large scale clusters (target up to 2000 nodes).
+* Single Node IPs (all traffic for a Service LB going through a single Node) is
```

Comment:
One of the mandate for upgrades is that there must be no disruption of services like APIserver or Ingress. Single Node IPs will probably fail to provide that right? the node will be drained and rebooted and then there will probably disruption of traffic while the IP moves to another node. https://github.com/openshift/enhancements/pull/356/files#diff-4efb0274cd0d2bdbac261b1c4f44cbacR46 points to 5 secs of fail over.

Is it possible to include no disruption of traffic when upgrading a goal? this would help target a specific case and not generic solution.

Reply:
When the apiserver goes down gracefully it notifies loadbalancers through the /readyz endpoint before it actually stops responding. We would probably want something similar for nodes being drained, a way to determine that it is happening so we can move the IP before it actually reboots and stops responding. That should avoid the failover delay because we'd be moving the IP from one still-functional node to another.

Reply:
There's a new upstream issue tracking the possibility of adding a more graceful failover procedure: https://github.com/metallb/metallb/issues/621

I'll also look at updating the requirements to reflect that we want graceful no-downtime failover for planned upgrades.  I think that will already be fine for the layer 3 (BGP) mode, but would take some more work for the layer 2 mode.